src/autopas/utils/Math.h

```diff
@@ -16,6 +17,165 @@ namespace autopas::utils::Math {
  */
 const double normalScale = 1. / std::sqrt(2 * M_PI);
 
+/**
+ * Addition function for integer types that is safe against over and underflow.
+ * If over or underflow is detected, the function returns the specified values.
+ * @tparam T
+ * @param a
+ * @param b
+ * @param valUnderflow Return value in case of underflow.
+ * @param valOverflow Return value in case of overflow.
+ * @return Sum or valOverflow or valUnderflow.
+ */
+template <class T, std::enable_if_t<std::is_integral_v<T>, bool> = true>
+T safeAdd(const T &a, const T &b, const T &valUnderflow = std::numeric_limits<T>::min(),
+          const T &valOverflow = std::numeric_limits<T>::max()) {
+  T result;
+  bool overflow = __builtin_add_overflow(a, b, &result);
+  if (overflow) {
+    // if both args are negative this is an underflow.
+    if (a < 0 and b < 0) {
+      result = valUnderflow;
+    } else {
+      result = valOverflow;
+    }
+  }
+  return result;
+}
+
+/**
+ * Addition function for floating point types that is safe against over and underflow.
+ * If over or underflow is detected, the function returns the specified values.
+ *
+ * @note Underflow here refers to a value more negative than representable and not the underflow gap around zero.
+ *
+ * @tparam T
+ * @param a
+ * @param b
+ * @param valUnderflow Return value in case of underflow.
+ * @param valOverflow Returns value in case of overflow.
+ * @return Sum or valUnderflow or valOverflow.
+ */
+template <class T, std::enable_if_t<std::is_floating_point_v<T>, bool> = true>
```

Comment:
would be so much nicer with c++20 Concepts/Constraints :P
```suggestion
template <std::floating_point T>
```
it only requires gcc10 or clang10 and libstdc++10 or libc++13 xD

maybe add a todo?